linkifiers: Add setting to edit linkifiers.

akshatdalton opened the pull request:
This commit adds a setting to edit realm filters. A new API is added in zproject/urls.py.Modal, added to edit realm filters, supports ui_report error.

Fixes #10830.

**Testing plan:** 
New test cases are added for this new setting.

**GIFs or screenshots:** 
#### Modal to Edit Linkifiers:
![Screenshot from 2021-01-12 21-31-31](https://user-images.githubusercontent.com/67277428/104345600-8e543900-5524-11eb-813a-64d249dffe4b.png)
-----------------------------------------------
#### Edit Linkifiers:
![Screenshot from 2021-01-12 21-33-30](https://user-images.githubusercontent.com/67277428/104345611-9318ed00-5524-11eb-834a-7c7e18d707e1.png)

timabbott commented on zerver/views/realm_filters.py:
```diff
@@ -41,3 +41,24 @@ def delete_filter(request: HttpRequest, user_profile: UserProfile, filter_id: in
+    url_format_string: str = REQ(),
+) -> HttpResponse:
+    try:
+        do_update_realm_filter(
```
We generally use `check_update_realm_filter` as the naming convention for functions where validation is done inside `actions.py`.  Though I notice this module isn't following that convention more generally, so it might be reasonable to defer to follow-up work.

akshatdalton replied:
I have left this unchanged, since you mentioned that this will be taken in the follow-up work.

timabbott commented on zerver/tests/test_realm_filters.py:
```diff
@@ -117,3 +117,71 @@ def test_delete(self) -> None:
+        data["url_format_string"] = "https://github.com/%(org)s/%(repo)s/issue/%(id)s"
+        result = self.client_patch(f"/json/realm/filters/{filter_id}", info=data)
+        self.assert_json_success(result)
+        self.assertIsNotNone(re.match(data["pattern"], "zulip/zulip#123"))
```
Are these tests basically a copy of the set of values from the "Create a linkifier" test?  Probably it's not worth duplicating them all, given that we know structurally we're running the exact same validation code in this code path.

akshatdalton replied:
I have reduced the number of test cases and have just put the basic test cases now.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -5851,6 +5851,51 @@ paths:
+        - name: filter_id
+          in: path
+          description: |
+            The ID of the filter that you want to update.
```
We use the term "linkifier" rather than "filter" in all user-facing docs, and probably this text should be included.

(More medium-term, we're going to eventually rename "filters" to "linkifiers" in the API; it's just a bit of an annoying transition because of mobile app compatibility issues)

akshatdalton replied:
I have renamed this to `linkifier` and have also changed the same over [here](https://github.com/zulip/zulip/pull/17045/files#diff-7d09c06f7ab16ed46eeff8eb6e1a342f75fc9f3c780c6f10643303145316581eR5895).

timabbott commented on zerver/lib/actions.py:
```diff
@@ -6451,6 +6451,17 @@ def do_remove_realm_filter(
+def do_update_realm_filter(realm: Realm, id: int, pattern: str, url_format_string: str) -> None:
+    pattern = pattern.strip()
+    url_format_string = url_format_string.strip()
+    realm_filter = RealmFilter.objects.get(realm=realm, pk=id)
```
Please use `id=id)`, rather than `pk=id`; we prefer that style as it's more readable.

akshatdalton replied:
It is corrected now. And I have also changed the same over [here](https://github.com/zulip/zulip/pull/17045/files#diff-9a1fd3f8a1bd0ccd686ac2d96d265d3d5dce323f6fc191adac2cb88500af6e1cR6444).

timabbott commented on static/js/settings_linkifiers.js:
```diff
@@ -105,6 +126,59 @@ exports.build_page = function () {
+        e.preventDefault();
+        e.stopPropagation();
+
+        const btn = $(this);
```
We prefer `e.currentTarget` and similar over `this` for this sort of selector, stylistically, though again I note this module seems to already have the style you're using.  Maybe a good follow-up project to bring this module in line with the more recently improved settings pages.

akshatdalton replied:
Again, I have left this unchanged, since you mentioned that this will be taken in the follow-up project.

timabbott commented on static/js/settings_linkifiers.js:
```diff
@@ -33,6 +36,24 @@ function sort_url(a, b) {
+    });
+    const div = $(html);
+    div.find("input").css("width", "400px");
+    div.find(".alert").css("margin-top", "10px");
```
These should be done via CSS rules -- wherever possible we avoid doing inline styles like this from JavaScript.

timabbott commented:
Thanks for working on this!  The UI looks pretty good (and actually, we probably want to think about replacing the existing ugly UI for adding a linkifier with a modal like that, ideally sharing most of the code.

Let me know once you've resolved the above comments! :)

timabbott commented:
I saw you did several pushes; just wanted to mention that it's important to specifically mention in GitHub when you're ready for another review, since otherwise one can't distinguish folks who are just pushing to run CI again or who think they've addressed everything.  Ideally, that comment should also discuss how you addressed the feedback.

akshatdalton commented:
Sorry @timabbott about this. I will keep this point in my mind.
I have updated the PR and have responded to your feedbacks.
Please review.

timabbott commented on zerver/views/realm_filters.py:
```diff
@@ -41,3 +41,24 @@ def delete_filter(request: HttpRequest, user_profile: UserProfile, filter_id: in
+        )
+        return json_success()
+    except ValidationError as e:
+        return json_error(e.messages[0], data={"errors": dict(e)})
```
This needs to catch RealmFilter.DoesNotExist.  You should use the same error message as for the deletion code path above.

akshatdalton replied:
But updating an existing filter is like adding a filter, hence errors each time can be different (cannot be always `RealmFilter.DoesNotExist`) like `Invalid filter pattern` (when new filter pattern entered is Invalid) or `Invalid URL format string` (when the new format string is Invalid).
Should I still update it to `RealmFilter.DoesNotExist` ?
Or should I add this as extra except block to catch `RealmFilter.DoesNotExist` ?

timabbott replied:
Yeah, I meant an additional except block -- sorry for the confusion.  I agree we still need the `except ValidationError` block that you already have.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -5895,6 +5895,51 @@ paths:
+      description: |
+        Update [linkifiers](/help/add-a-custom-linkification-filter), regular
+        expression patterns that are automatically linkified when they appear
+        in messages and topics.
```
This should make it more clear that this is updating a single linkifier.

akshatdalton replied:
Done

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -5895,6 +5895,51 @@ paths:
+          schema:
+            type: string
+          example: https://github.com/zulip/zulip/issues/%(id)s
+          required: true
```
For these last two parameters, since they should match what's present with the "add realm filter" endpoint (and I assume you just copied), it's worth extracting this as a shared variable.

akshatdalton replied:
Sorry, I didn't get what `extracting this as a shared variable` means here.

timabbott replied:
We have a set of common components in the API docs like `#/components/schemas/BasicStream` that let us avoid duplicating code in the OpenAPI files -- we should use that mechanism for this as well.

timabbott commented on static/templates/admin_linkifier_edit_form.hbs:
```diff
@@ -0,0 +1,27 @@
+        <div class="alert-notification edit-filter-status"></div>
+    </div>
+    <div class="modal-body">
+        <div id="linkifier-form" data-filter-id="{{filter_id}}">
```
Might as well call this `data-linkifier-id` just to avoid a migration later.

akshatdalton replied:
Done.

timabbott commented on static/styles/settings.css:
```diff
@@ -1873,3 +1873,13 @@ body:not(.night-mode) #settings_page .custom_user_field .datepicker {
+#edit-filter-pattern-status,
+#edit-filter-format-status {
+    margin-top: 10px;
+}
```
Same with these; s/filter/linkifier` would be well worth doing in this PR just to avoid future churn -- we're planning to eliminate `realm_filters` from the lexicon.

Also, I'd probably prefer to not mix `_`s and `-`s for naming IDs for the same feature.

timabbott commented on static/js/settings_linkifiers.js:
```diff
@@ -109,6 +128,59 @@ export function build_page() {
+                    overlays.close_modal("#linkifier-info-form-modal");
+                },
+                error_continuation(xhr) {
+                    change_filter_button.prop("disabled", false);
```
Maybe worth adding a brief comment here that the endpoint uses the somewhat uncommon in our codebase Django ValidationError system that returns somewhat complicated error dictionaries.

This isn't a pattern we want to use in general, but it's probably OK to have it here given that the other linkifier endpoints have it too.

That said, it'd be nice to have a prep commit that extracts the code here that's common with the "add linkifier" code path as a function that we can call from both places rather than copying it.

In general, we try to avoid copying substantial blocks of code, as it tends to result in someone later fixing bugs in only one of the two places.

akshatdalton replied:
Done.

timabbott commented:
@akshatdalton it looks like you haven't pushed your last batch of changes to this PR?  The most recent push is before my last batch of comments.

akshatdalton commented:
@timabbott sure I'll prioritize this one.
Actually, earlier I needed your feedbacks on my comment to proceed forward.
Before my next push, can you please answer to [here](https://github.com/zulip/zulip/pull/17045#discussion_r596079248)?

timabbott commented:
Already done, I think :)

akshatdalton commented:
Following are the points I have addressed in my current `push`:

- As a prep commit, I have added the helper function to handle API errors as asked [here](https://github.com/zulip/zulip/pull/17045#discussion_r594051441).
- In `zulip.yaml`, realm `pattern` and `url_format_string` are extracted as `Shared Paramter` as suggested [here](https://github.com/zulip/zulip/pull/17045#discussion_r594050432).
- Made clear in `zulip.yaml` that the added API updates one linkifier at a time. https://github.com/zulip/zulip/pull/17045#discussion_r594050188
- Added an extra except block for `RealmFilter.DoesNotExist` error. https://github.com/zulip/zulip/pull/17045#discussion_r601143787, a test case for the same is added as well.
- Solved issue regarding using the term `linkifier` rather than `filter` is addressed as well.

In `zulip.yaml`, `filter_id` is also used at several places (like delete linkifier and update linkifier), should I also extract them as `Shared Parameter`?

@timabbott please review.

akshatdalton commented:
## Screenshot showing UI error report handling for `Filter not found` error.
![Screenshot from 2021-03-26 00-39-25](https://user-images.githubusercontent.com/67277428/112529934-db2d7900-8dcb-11eb-84b8-1a734e2269a6.png)

akshatdalton commented:
I haven't added `puppeteer_test` for this new setting.
Working on that.

akshatdalton commented:
This might be delayed a little bit. The puppeteer test to edit realm-linkifier is failing in a strange case.
Related discussion in: https://chat.zulip.org/#narrow/stream/43-automated-testing/topic/puppeteer.20issue

akshatdalton commented:
I have added the Puppeteer test; I think it covers all the possible cases.

[Above](https://github.com/zulip/zulip/pull/17045#issuecomment-807300401) I have posted the comment describing what the new changes are.
@timabbott, please review.

akshatdalton commented:
Okay, I wanted to resolve a small conflict through GitHub only but did want this to happen: `Merge branch 'master' into issue_#10830`.

akshatdalton commented:
Just resolved the conflicts!

timabbott commented:
Sorry I missed this before we merged changes making a bunch more conflicts; can you rebase one more time?

akshatdalton commented:
To summarise once again, following are the points I have addressed in my current `push`:

* As a prep commit, I have added the helper function to handle API errors as asked [here](https://github.com/zulip/zulip/pull/17045#discussion_r594051441).
* In `zulip.yaml`, realm `pattern` and `url_format_string` are extracted as `Shared Paramter` as suggested [here](https://github.com/zulip/zulip/pull/17045#discussion_r594050432).
* Made clear in `zulip.yaml` that the added API updates one linkifier at a time. [#17045 (comment)](https://github.com/zulip/zulip/pull/17045#discussion_r594050188)
* Added an extra except block for `RealmFilter.DoesNotExist` error. [#17045 (comment)](https://github.com/zulip/zulip/pull/17045#discussion_r601143787), a test case for the same is added as well.
* Solved issue regarding using the term `linkifier` rather than `filter` is addressed as well.

In `zulip.yaml`, `filter_id` is also used at several places (like delete linkifier and update linkifier), should I also extract them as `Shared Parameter`?

@timabbott please review.

abhijeetbodas2001 commented on zerver/openapi/zulip.yaml:
```diff
@@ -6589,11 +6571,38 @@ paths:
+      operationId: update_linkifier
+      tags: ["server_and_organizations"]
+      description: |
+        Update [linkifiers](/help/add-a-custom-linkification-filter), regular
```
This should be `/help/add-a-custom-linkifier`. This page was renamed in 9223dced3b1f0ca28cf9cf86193e121167ab50ef.

abhijeetbodas2001 commented on zerver/openapi/zulip.yaml:
```diff
@@ -11506,3 +11515,25 @@ components:
+        type: string
+      example: "#(?P<id>[0-9]+)"
+      required: true
+    LinkifierURLFormatString:
```
Would it be better to do this extraction in a separate, prep commit?

abhijeetbodas2001 commented on zerver/views/realm_linkifiers.py:
```diff
@@ -41,5 +41,28 @@ def delete_linkifier(
         do_remove_linkifier(realm=user_profile.realm, id=filter_id)
     except RealmFilter.DoesNotExist:
-        return json_error(_("Filter not found"))
+        return json_error(_("Filter not found."))
```
This should be `linkifier`, not `filter`. I think I missed this when doing the renaming for most of these 😅. Can you make a separate commit for this, along with the punctuation fix?

abhijeetbodas2001 commented on zerver/views/realm_linkifiers.py:
```diff
@@ -41,5 +41,28 @@ def delete_linkifier(
+        )
+        return json_success()
+    except RealmFilter.DoesNotExist:
+        return json_error(_("Filter not found."))
```
This should also be `linkifier`.

abhijeetbodas2001 reviewed (commented):
Thanks for working on this! I've gone through the code and left a few mostly minor comments on naming etc.

akshatdalton commented:
@abhijeetbodas2001, thanks for the feedback! I'll update the PR as soon as possible.

akshatdalton commented:
Based on the feedback, I've extracted the shared parameter and renamed `Filter` to `Linkifier` as separate commits.

akshatdalton commented:
@timabbott a gentle reminder about this PR.

timabbott commented:
@abhijeetbodas2001 do you have time to review this one?  You have a lot of context after the recent data format PR.

@akshatdalton you should also fix CI.

timabbott commented:
@akshatdalton sorry to make you do this again so soon, but this needs a rebase for the i18n library change.  

(Also, CI is failing).

akshatdalton commented:
@timabbott, I just resolved the conflicts and CI issue. Hopefully, this will be pass now.

akshatdalton commented:
Okay, everything is resolved up till now.

Whenever modal opens to edit the realm linkifiers, `pattern` and `url_format` values are already present in the input text field.
This case is not covered in the puppeteer test cases that I have added. Should I go ahead and add the related test? Or it doesn't seem that important?

abhijeetbodas2001 commented:
I'm a bit busy with college assignments today. Will definitely try to review and test this tomorrow.

timabbott commented:
I don't think it's important to have complete puppeteer tests for settings UI.

timabbott commented:
I merged the first two commits as the series ending with 53f123df72c95bf181428a4fd6ade0ee70bd2695.

timabbott commented on zerver/tests/test_realm_linkifiers.py:
```diff
@@ -109,8 +109,55 @@ def test_delete(self) -> None:
+        result = self.client_get("/json/realm/linkifiers")
+        self.assert_json_success(result)
+        self.assertEqual(200, result.status_code)
+        self.assertEqual(len(result.json()["linkifiers"]), 1)
```
You should verify the value returned here, to make sure it was actually updated.

timabbott commented:
Can you extract the backend/API changes as a separate commit from the frontend? It should be mergeable independently, and that'll allow us to avoid merge conflicts for it.

Also, you need to add the new endpoint to the API changelog (templates/zerve/rapi/changelog), bump API_FEATURE_LEVEL, and add a `**Changes**` entry to the new endpoint in `zulip.yaml`.

Other than that, the backend looks good to me.

abhijeetbodas2001 commented on zerver/views/realm_linkifiers.py:
```diff
@@ -43,3 +43,26 @@ def delete_linkifier(
+        )
+        return json_success()
+    except RealmFilter.DoesNotExist:
+        return json_error(_("Filter not found"))
```
This should be `Linkifier not found` 😅

abhijeetbodas2001 replied:
Oh, I see you've fixed this in a later commit in this PR. But why introduce it in the first place?

abhijeetbodas2001 replied:
Yeah, my point was, you can change just the `delete_linkifier` code in a commit at the beginning of this PR, and then just write this new code correctly.

abhijeetbodas2001 commented on zerver/openapi/python_examples.py:
```diff
@@ -372,6 +372,29 @@ def add_realm_filter(client: Client) -> None:
+
+    # Test `Filter not found` error
+    result = client.call_endpoint(url=f"/realm/filters/{filter_id}", method="PATCH")
+    validate_against_openapi_schema(result, "/realm/filters/{filter_id}", "patch", "400")
```
I do not understand how using the same linkifier ID gives a `200` response at one time and a `400: Filter not found` at another time. Do we or do we not have a linkifier with `id=1` in the database?
I think we don't need to test bad inputs here, because the point of this file is to generate correct examples.
You can probably create a new linkifier outside the `code_example` block at the beginning of the test, and then use the ID you get there to update.

akshatdalton replied:
> I do not understand how using the same linkifier ID gives a 200 response at one time and a `400: Filter not found` at another time.

That's because in the 2nd call to `client.call_endpoint` I'm not passing `request=request`. But I guess I confused here by not doing `{filter_id + 1}`. Sorry for the confusion.

abhijeetbodas2001 replied:
Ok, that makes sense. But I still think there's no need to test the `Filter not found` case, because it doesn't add to the documentation.

akshatdalton replied:
Cool! I should better remove that part. Also, since `*_realm_filter` functions do not do the same.

abhijeetbodas2001 commented on zerver/openapi/zulip.yaml:
```diff
@@ -6659,7 +6659,7 @@ paths:
           in: path
           description: |
-            The ID of the filter that you want to remove.
+            The ID of the linkifier that you want to remove.
```
This could be in a separate commit, because it is unrelated with this new feature.

abhijeetbodas2001 reviewed (commented):
I left some more comments after going through the backend code. Also, tests should be in the same commit which introduces new code, so the first two commits here should be squashed. (And I think so can be the puppeteer tests)
One more thing is that you use the term "update" throughout this. I personally feel "edit" feels much more intuitive than "update", but maybe it doesn't matter.

akshatdalton commented:
> I left some more comments after going through the backend code. Also, tests should be in the same commit which introduces new code, so the first two commits here should be squashed. (And I think so can be the puppeteer tests)
> One more thing is that you use the term "update" throughout this. I personally feel "edit" feels much more intuitive than "update", but maybe it doesn't matter.

Ah!, I had a feeling that tests should be in the same commit. Thanks for confirming this! :+1:

akshatdalton commented on zerver/openapi/zulip.yaml:
```diff
@@ -6671,6 +6671,35 @@ paths:
+      description: |
+        Update [linkifiers](/help/add-a-custom-linkifier), regular
+        expression patterns that are automatically linkified when they appear
+        in messages and topics. This update one linkifier at a time.
```
Yeah!, actually I was following the documentation description of other APIs. I saw that `delete/remove linkifier` also has `Update [linkifiers]...` so I thought to go with the current description. But thanks, I guess this makes more sense.

akshatdalton commented on zerver/tests/test_realm_linkifiers.py:
```diff
@@ -12,8 +12,10 @@ def test_list(self) -> None:
         self.assert_json_success(result)
-        self.assertEqual(200, result.status_code)
-        self.assertEqual(len(result.json()["linkifiers"]), 1)
+        linkifier = result.json()["linkifiers"]
+        self.assertEqual(len(linkifier), 1)
+        self.assertEqual(linkifier[0]["pattern"], "#(?P<id>[123])")
+        self.assertEqual(linkifier[0]["url_format"], "https://realm.com/my_realm_filter/%(id)s")
```
I think I should do:
```.py
linkifiers = result.json()["linkifiers"]
self.assertEqual(len(linkifiers), 1)
self.assertEqual(linkifiers[0]["pattern"], "#(?P<id>[123])")
self.assertEqual(linkifiers[0]["url_format"], "https://realm.com/my_realm_filter/%(id)s")
```
`linkifiers` would make more sense.

akshatdalton commented on zerver/tests/test_realm_linkifiers.py:
```diff
@@ -114,3 +116,55 @@ def test_delete(self) -> None:
+        # Verify that the linkifier is updated accordingly.
+        linkifier = result.json()["linkifiers"]
+        self.assertEqual(len(linkifier), 1)
+        self.assertEqual(linkifier[0]["pattern"], "#(?P<id>[0-9]+)")
```
https://github.com/zulip/zulip/pull/17045#discussion_r614111881

akshatdalton commented on zerver/tests/test_realm_linkifiers.py:
```diff
@@ -114,3 +116,55 @@ def test_delete(self) -> None:
+        do_update_linkifier(
+            realm,
+            linkifier_id,
+            "#(?P<id>[0-9]+)",
```
Again, I was trying to follow `test_delete` and `test_list` functions. But I think this makes more sense. I'll update accordingly in my new commit and address the older changes in a separate commit.

akshatdalton commented on zerver/tests/test_realm_linkifiers.py:
```diff
@@ -114,3 +116,55 @@ def test_delete(self) -> None:
+        }
+        result = self.client_patch(f"/json/realm/filters/{linkifier_id}", info=data)
+        self.assert_json_error(
+            result, "Invalid linkifier pattern.  Valid characters are [ a-zA-Z_#=/:+!-]."
```
These validations are done since a new API is added (hence I've used `client_patch`).
All the validations/error validations are done to verify that they work correctly for the new added API.
I've not repeated any validation like `create` tests; here, I've just made sure that we do all types of validations (via this new API).

akshatdalton commented on zerver/tests/test_realm_linkifiers.py:
```diff
@@ -114,3 +116,55 @@ def test_delete(self) -> None:
+        ] = "https://example.com/~user/web#view_type=type&model=model&action=12345&id=%(id)s"
+        result = self.client_patch(f"/json/realm/filters/{linkifier_id}", info=data)
+        self.assert_json_success(result)
+        self.assertIsNotNone(re.match(data["pattern"], "PR 123"))
```
Doing an `assert_json_success` implies that the request was made successfully or not.
`assertIsNotNone` is done to verify that the `notify_linkifiers` function (in `actions.py`) sends an event containing the updated linkifiers that we cross-check in `assertIsNotNone`.

abhijeetbodas2001 replied:
Sorry for the confusion here. What I meant was that, once you have addressed https://github.com/zulip/zulip/pull/17045#discussion_r613996007, you can club that and this together, because they both test a successful linkifier update.

akshatdalton replied:
Got it! That will help to reduce these duplicate tests when https://github.com/zulip/zulip/pull/17045#discussion_r613996007 is addressed. Thanks!

akshatdalton commented:
The following points are addressed in my current push:

- I've squashed the commits containing new tests. https://github.com/zulip/zulip/pull/17045#pullrequestreview-636560606
- `pk` -> `id` cleanup is done in a separate commit. https://github.com/zulip/zulip/pull/17045#discussion_r613976081
- Updated OpenAPI documentation so that the description to update the linkifier is more concise. https://github.com/zulip/zulip/pull/17045#discussion_r613987230
- Replaced `linkifier` with `linkifiers` in `test_realm_linkifiers.py`; since the latter one makes more sense. https://github.com/zulip/zulip/pull/17045#discussion_r614111881
- Used backend-endpoint function instead of `do_*` helper functions in `test_realm_linkifiers.py` so that tests are more end-to-end. https://github.com/zulip/zulip/pull/17045#discussion_r614131114 & https://github.com/zulip/zulip/pull/17045#discussion_r614157955
- Renamed `Filter` to `Linkifier` in the first commit. https://github.com/zulip/zulip/pull/17045#discussion_r614124183
- Initially, I pushed a commit extracting `LinkifierID` as a Shared Parameter in `zulip.yaml` but found that its description differs for each endpoint, so I have not extracted it.

@timabbott, @abhijeetbodas2001, please review.

abhijeetbodas2001 reviewed (commented):
The backend LGTM! I also tested this out manually, and mostly works as expected. Some questions I have on the frontend are-
- If you do this-
   - Open two tabs, both at `/linkifier-settings`
   - Open the edit modal for a linkifier in one tab
   - Delete that linkifier from the other tab (while the edit modal is still open in the first one)
   - Edit and save the linkifier in the first tab.
   
   Then, you get an error like this-
   ![image](https://user-images.githubusercontent.com/55339528/114971105-51883d00-9e99-11eb-912a-1ff243c5084d.png)
   Which is correct, but I think in such a situation, we should close the edit modal, since it contains bad/outdated data. I think showing the error only in the main settings page should be enough (especially for this `not found` error).

- After successfully saving a linkifier, we should either clear the `filter` box in the settings page, or refresh the list, because the filter could still incorrectly match to the old data like so-
   
    ![image](https://user-images.githubusercontent.com/55339528/114971675-78933e80-9e9a-11eb-9e7d-ef537b06aeb0.png)
    What I did here was, made a linkifier to `zulip-mobile` repository, then typed in `mobile` in the search box, and the edited the linkifier to be just `zulip`.

- When providing bad inputs (empty field for example), we get an error message like so-
    ![image](https://user-images.githubusercontent.com/55339528/114971889-eb9cb500-9e9a-11eb-986d-e7ae062f39b0.png)
   Which is nice. The problem is, the error message is shown on the main settings page even after closing the edit modal-
   ![image](https://user-images.githubusercontent.com/55339528/114971969-16870900-9e9b-11eb-8cec-19a1c3813e44.png)
   without any context. I think it would be OK to show these errors just in the edit modal, and not in the main settings page.

akshatdalton commented:
Yeah, the first two seems to be contrived since any change in settings of Linkifiers is done by the admin, so they will rarely get any such case.

For the last one, the error is shown in the main modal also since it may happen that the user tried to change the linkifiers and some error occurred. And right after completion of the process, they closed the modal, so they might want to know the reason for the failure that they can see on the main modal itself. In the other case, such notification can also help the user to know if their change was successful or not after completion of the process.

timabbott commented:
I merged the backend commit series ending with 6509c4f8f4b816ba80313ca0ffb8f8c306d56a85, thanks @akshatdalton!

I'll post a batch of comments on the frontend.

timabbott commented on frontend_tests/puppeteer_tests/realm-linkifier.ts:
```diff
@@ -54,14 +54,96 @@ async function test_invalid_linkifier_pattern(page: Page): Promise<void> {
+    await page.waitForSelector("#linkifier-edit-form-modal", {hidden: true});
+    await page.waitForFunction(() => $(".edit-linkifier-status").text().trim() === "Saved");
+    await page.waitForSelector(".linkifier_row", {visible: true});
+    await page.waitForFunction(() => $("#admin_linkifiers_table").attr("data-status") === "saved");
```
What is this `data-status` attribute for -- just for tests?

timabbott commented on static/js/settings_linkifiers.js:
```diff
@@ -40,6 +43,22 @@ function sort_url(a, b) {
+        pattern: linkifier.pattern,
+        url_format_string: linkifier.url_format,
+    });
+    const div = $(html);
```
Let's rename `html` to `edit_form_html` and perhaps just not have a `div` local variable, doing the `$(edit_form_html)` below instead.

timabbott commented on static/js/settings_linkifiers.js:
```diff
@@ -40,6 +43,22 @@ function sort_url(a, b) {
+    modal_container.empty().append(div);
+    overlays.open_modal("#linkifier-edit-form-modal");
+
+    return {modal: div};
```
This is kinda a weird return format; why not just return `div`?

timabbott commented on static/js/settings_linkifiers.js:
```diff
@@ -65,6 +84,7 @@ export function populate_linkifiers(linkifiers_data) {
     }
 
     const linkifiers_table = $("#admin_linkifiers_table").expectOne();
+    $(linkifiers_table).attr("data-status", "saved");
```
What's the story with this `data-status` attribute?  We don't AFAIK use similar attributes anywhere except our standard `settings_save_discard_widget.hbs` widget; are you trying to replace its UI?

timabbott commented on static/templates/admin_linkifier_list.hbs:
```diff
@@ -11,6 +11,9 @@
             <i class="fa fa-trash-o" aria-hidden="true"></i>
         </button>
+        <button class="button small edit btn-warning" data-linkifier-id="{{id}}">
+            <i class="fa fa-pencil" aria-hidden="true"></i>
```
This should be `aria-label` (with a title and tippy tooltip), not `aria-hidden`, sine the button does not have a textual label.

(`aria-hidden` is only for icons on buttons that have a plain-text label too, e.g. in our gear menu).

timabbott replied:
(The adjacent button is broken the same way; can you fix that in a prep commit?)

timabbott commented:
I think we're getting close to being able to merge this; the main thing we'll need is to figure out how to avoid that `data-status` puppeteer testing hack, if we can.  

@Riken-Shah FYI in case you can help with the puppeteer piece here.

Riken-Shah commented:
@akshatdalton How about if we use `#linkifier-field-status` to check if it's saved or not?

akshatdalton commented:
> How about if we use `#linkifier-field-status` to check if it's saved or not?

@Riken-Shah, that won't help. [Here](https://github.com/zulip/zulip/pull/17045/files#diff-50608ad25da94bcf381a5a5f64df455d01948a407abe695d0828d4244a6d6ce1R171-R175) is the note that I added, the reason behind adding `data-status`.
`data-status` how it works that (and an alternative that I need) **:** Puppeteer immediately looks for the updated value in the table row (right after modal is closed). After making a `PATCH` request, the `success` method updates the UI (including that `#linkifier-field-status`). But during that moment, the table row doesn't update immediately. `send_event` (in the `notify_linkifiers` function) fires an event, sending the new realm linkifiers list, which then updates the UI (table rows). So, this round update is making it harder to check the UI updates. And that's my motive to add `data-status`; I set its attribute to `unsaved` once the `submit` button is clicked. and it is updated to `saved` once we have [got the updated](https://github.com/zulip/zulip/pull/17045/files#diff-50608ad25da94bcf381a5a5f64df455d01948a407abe695d0828d4244a6d6ce1R87) list of realm linkfiers.

akshatdalton commented:
Hmm, so I was thinking how the [HACK](https://github.com/zulip/zulip/commit/f2c9d909e32aac6d7f306a39a614dcd5b834089a#diff-50608ad25da94bcf381a5a5f64df455d01948a407abe695d0828d4244a6d6ce1R171-R175) can be avoided. So, I looked at how `delete linkifier` is implemented in the Puppeteer test. Then I noticed that it removes the table row as soon as the request made was successful. And hence, in `edit linkifier`, I have done the same thing, i.e. as soon as the request made was successful, I update the table row on the frontend side. And this can't raise any ambiguity since the request made was successful; this implies our operation: `delete/edit` was successful. And hence, we can safely update the same on the frontend side.

Though there are still some issues that I still face with Puppeteer tests, e.g., I have changed the [order in which functions](https://github.com/zulip/zulip/pull/17045/files#diff-2499f56e02b20f708883a5c5bdd41f64fc24d3ef727fc3a4b199b95f43f4310bR143-L64) are called. This is OK, as it is not counted as some HACK, I guess, and neither affects the other tests. Still, I can't understand that if I maintain the [previous order](https://github.com/zulip/zulip/pull/17045/files#diff-2499f56e02b20f708883a5c5bdd41f64fc24d3ef727fc3a4b199b95f43f4310bL63-L64) in which the functions are called; the linkifier row is not deleted. :man_shrugging:.

akshatdalton commented:
Following are the points, I have addressed in my current push:

- Removed the `data-status` attribute. https://github.com/zulip/zulip/pull/17045#discussion_r616273201
- Renamed `html` => `edit_form_html` and `div` => `edit_form_div`. https://github.com/zulip/zulip/pull/17045#discussion_r616273770
- Improved the weird return format. https://github.com/zulip/zulip/pull/17045#discussion_r616273887. Initially, I was trying to follow other code paths, I guess :thinking:.
- Added `aria-label` attribute and removed `aria-hidden` attribute. https://github.com/zulip/zulip/pull/17045#discussion_r616275251
- Fixed the `aria-label` attribute of the `Delete` button in the prep commit. https://github.com/zulip/zulip/pull/17045#discussion_r616275340

@timabbott, please review.

Riken-Shah commented:
@akshatdalton How about this?

```javascript
await page.waitForFunction(() => $(".linkifier_pattern").text().indexOf(pattern) !== -1);
```
We can also go with, 
```javascript
await page.waitForFunction(() => $(".linkifier_pattern").text() === pattern);
```
but it will not work if there is more than one entry and ours is the last one.

akshatdalton commented:
@Riken-Shah Thanks for helping me out!
I think my current changes are in sync with `delete linkifier` and I have also given [my explanation](https://github.com/zulip/zulip/pull/17045#issuecomment-823647163) that why they make sense.
So, I think they will be better.

But for future purposes can you investigate why [changing the order of function](https://github.com/zulip/zulip/pull/17045/files#diff-2499f56e02b20f708883a5c5bdd41f64fc24d3ef727fc3a4b199b95f43f4310bR143-L64) makes the Puppeteer test fail? (Related https://github.com/zulip/zulip/pull/17045#issuecomment-823647163)

akshatdalton commented:
Just resolved the conflicts!

I have addressed the new changes from [here](https://github.com/zulip/zulip/pull/17045#issuecomment-823647163).
@timabbott, a gentle reminder about this PR.

timabbott commented on static/templates/settings/admin_linkifier_list.hbs:
```diff
@@ -9,7 +9,7 @@
     <td class="no-select actions">
         <button class="button small delete btn-danger" data-linkifier-id="{{id}}">
-            <i class="fa fa-trash-o" aria-hidden="true"></i>
+            <i class="fa fa-trash-o" data-tippy-content="{{t "Delete" }}" aria-label="{{t "Delete" }}" title="{{t "Delete" }}"></i>
```
You don't need both `title` and a tooltip; additionally, when I hover over this button, nothing happens.

It's really important to test even tiny changes like this, because even though it's very simple, a rule of software is that essentially nothing works that you haven't tested :)

akshatdalton replied:
From https://github.com/zulip/zulip/pull/17045#discussion_r616275251, I assumed that we need both `title` and `data-tippy-content` attribute.
And sorry for the result, actually, I cross-checked the changes and it was working fine for me; the reason that why it was not showing to you was that: since I added the `title` attribute to the `fa fa` icon and not to the `button` therefore it needed hovering exactly over the icon (hovering over the button and not over the icon doesn't give the result).

akshatdalton commented:
Now, I have only added the `title` attribute to the `button`. (Related new comment: https://github.com/zulip/zulip/pull/17045#discussion_r621898609).
The result of the changes are similar to the section: **Settings** > **Users**.
The thing that I mentioned in https://github.com/zulip/zulip/pull/17045#discussion_r621898609 can be observed in the section: **Settings** > **User Groups** (i.e. hovering over the button doesn't show any result, but if we hover over the icon specifically then we can see the tooltip). (And this probably needs to be corrected if that's the case)

In the comment, where you mentioned in https://github.com/zulip/zulip/pull/17045#discussion_r616275251 about the `aria-hidden` attribute -- this is also present in the section: **Settings** > **Users** (which probably needs to be corrected). 

I have checked the changes, and they work fine now.
@timabbott, please review.

timabbott commented on static/templates/settings/admin_linkifier_list.hbs:
```diff
@@ -11,6 +11,9 @@
         <button class="button small delete btn-danger" data-linkifier-id="{{id}}" title="{{t "Delete" }}">
             <i class="fa fa-trash-o"></i>
         </button>
+        <button class="button small edit btn-warning" data-linkifier-id="{{id}}" title="{{t "Edit" }}">
```
I fixed several issues here:
* Missing aria-label, which should always be present alongside title.
* Invalid quoting within the `t` tag -- you need to use a different quote type.
* The edit button should be before the delete button.

I also notice there's a weird margin within the pencil icons we need to fix; will attempt a new commit to fix that.

andersk replied:
Although the quoting looks weird, it isn’t actually a problem. `{{t "Edit" }}` is evaluated by Handlebars well before the outer quotes are parsed as HTML.

timabbott replied:
Yeah, but it confuses one's editor (etc.), and thus seemed worth fixing.

timabbott commented on static/templates/settings/admin_linkifier_edit_form.hbs:
```diff
@@ -0,0 +1,27 @@
+                    <div class="alert" id="edit-linkifier-pattern-status"></div>
+                </div>
+                <div class="input-group name_change_container">
+                    <label for="linkifier-url-format-string" >{{t "URL format string" }}</label>
```
The label `for` element needs to point to the ID it goes with.

timabbott commented on static/js/settings_linkifiers.js:
```diff
@@ -127,6 +146,61 @@ export function build_page() {
+                success_continuation() {
+                    const row = btn.parents("tr");
+                    row.find(".linkifier_pattern").val(pattern);
+                    row.find(".linkifier_url_format_string").val(url_format_string);
```
These row edit changes are unnecessary; we'll live-update the row via `server_events_dispatch` like we do other UI.  I removed them and verified the UI still works.

timabbott commented:
I merged this after the changes noted above and adding a couple more commits of my own, thanks @akshatdalton!

This PR had an unusually large amount of back-and-forth.  Please try to take lessons from this in thinking through what each detail means, and making less use of copy-paste.
